### DISCUSSION TITLE
Stop HTTP Sender Send Attempts

### DISCUSSION THREAD

**Discussion by JJWest12** [nextgenhealthcare/connect#6002] [Q&A]
I am trying to work on a channel that performs a GET API call. This API call will often return an error response stating that there is no data in the time range given. When this response happens it isn't necessarily an "error" however. The destination has 3 retries, so I am trying to cut it off to not do additional retry attempts if that specific errors is returned on the first attempt. No matter what I have tried it is always doing the 3 retires before throwing my error.

I have been searching around these forums but cannot seem to get the results I am looking for. I saw something about having the destination response transformer inbound set to Raw and even that is not helping. Below is the code I have in my response transformer, am I missing something obvious here? Or is this even possible to do?

    var apiResponse = response.getMessage()

    if(apiResponse.includes('No documents that meet the'){
        //DO NOT DO ADDITIONAL RETRIES
        throw 'DO NOT DO ADDITIONAL RETRIES'
    }

**Comment by jonbartels** [comment]
Look at this example that limits send attempts: https://appsbyaaron.wordpress.com/2017/11/01/force-a-queued-message-to-error-if-the-number-of-send-attempts-exceeds-a-threshold/

You want to set the message state, not just throw. The throw causes the connector to think it just errored as if it were a timeout or processing issue. Setting the message state causes the message to drop out of the queue.

**Reply by pacmano1** [reply]
To fully achieve this, typically you want to set the response data type to raw and also serialize (or JSON.parse) it in a try..catch.  You both then check the response code (ie "200") and the quality of the response itself, setting it to queued, sent or errored.  This would normally involve manually controlling the send attempts.

**Reply by JJWest12** [reply]
I am doing the check in the response transformer, but still see it could through 4 send attempts (the original plus 3 retries)

    if(responseStatus == ERROR    ){
	    responseStatus = SENT   								
    }

**Reply by pacmano1** [reply]
I do something like this - but note response data type is set to RAW and I never set the respone status to ERROR, instead mark it as a sent and call a function for further actions.

I would probably use $co here not $c if multiple destinations.   Destination is set to always queue in this use case.

```JS
try {
    msg = JSON.parse(msg)             // we are going to reset msg to be its value in JSON here.
    $c('isJSONresponse', true)
} catch (e) {                         // we have an empty msg or some other non-JSON response 
    $c('isJSONresponse', false)
}

var responseStatusLine = $('responseStatusLine');
var responseCode = parseInt(responseStatusLine.split(' ')[1], 10);

if ( responseCode == 200) {    // we have an entirely successful message (note no check on isJSON, buyer beware)
        $c('API_Post_OK',true)
        process_results();
} else {                                    // we either have an entirely bad post (endpoint not answering OR we have a JSON error)
    if ($c('isJSONresponse') == true) {     // do not retry, just collect the JSON stuff
        responseStatus = SENT
        bad_call_error_handler(JSON.stringify(msg))
    } else {                                    // Let's retry a few times
        if (connectorMessage.getSendAttempts() <= 2) { // try a few times
            responseStatus = QUEUED                 // 
        } else {                                // exceeded the tries
              responseStatus = SENT
              $c('API_Post_OK',false)
              bad_call_error_handler(msg)                    // we have tried this post a few times, responseStatus will already be ERROR, so just send the alert
        }
    }
}
```

**Reply by JJWest12** [reply]
I have the datatype to raw and it is still running through all 4 attempts before ultimately setting the destination to say "SENT". I am just not sure what I am missing here

**Reply by pacmano1** [reply]
@JJWest12 you just have to set the responseStatus to SENT for your conditon.   Maybe post your code?

**Reply by tonygermano** [reply]
You probably have your destination set to queue on failure and the number of retries before queuing set to 3. Change the number of retries to 0 because you are handling the number of retries in the response transformer.
